express-notes-api/index.js

```diff
@@ -0,0 +1,148 @@
+const express = require('express');
+const fs = require('fs');
+const app = express();
+const data = require('./data.json');
+const notes = data.notes;
+const notesArray = [];
+const nextId = data.nextId;
+const resError = {};
+
+for (const key in notes) {
+  notesArray.push(notes[key]);
+}
+
+app.get('/api/notes', (req, res) => {
+  res.send(notesArray);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const id = parseInt(req.params.id);
+  if (id > 0) {
+    let index = -1;
+    for (let i = 0; i < notesArray.length; i++) {
+      if (notesArray[i].id === id) {
+        index = i;
+        break;
+      }
+    }
+    if (index === -1) {
+      resError.error = `cannot find note with id ${id}`;
+      res.status(404).json(resError);
+    } else {
+      res.status(200).send(notesArray[index]);
+    }
+  } else {
+    resError.error = 'id must be a positive integer';
+    res.status(400).json(resError);
+  }
+});
+
+app.use(express.json());
+
+app.post('/api/notes', (req, res) => {
+  const newNote = req.body;
+  for (const key in newNote) {
+    if (key !== 'content') {
+      resError.error = 'content is a required field';
+      res.status(400).json(resError);
+    } else {
+      newNote.id = nextId;
+      notes[nextId] = newNote;
+      data.nextId = newNote.id + 1;
+      const dataJson = JSON.stringify(data, null, 2);
+      fs.writeFile('data.json', dataJson, 'utf8', err => {
+        if (err) {
+          resError.error = 'An unexpected error occurred.';
+          res.status(500).json(resError);
+        }
+      });
+      res.status(200).send(newNote);
+    }
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const id = parseInt(req.params.id);
+  if (id > 0) {
+    for (const key in notes) {
+      if (parseInt(key) === id) {
+        delete notes[key];
+        res.sendStatus(204);
+        const dataJson = JSON.stringify(data, null, 2);
+        fs.writeFile('data.json', dataJson, 'utf8', err => {
+          if (err) {
+            resError.error = 'An unexpected error occurred.';
+            res.status(500).json(resError);
+          }
+        });
+      }
+    }
+    resError.error = `cannot find note with id ${id}`;
+    res.status(404).json(resError);
+  } else {
+    resError.error = 'id must be a positive integer';
+    res.status(400).json(resError);
```

Comment:
Try writing a few conditionals up front.
- if the id is invalid, send a `400`
- if the note doesn't exist, send a `404`
- delete the note and write the file
  - if writing fails, send a `500`
  - if writing succeeds, send a `204`